### ISSUE TITLE
Top-level @var ignored if file has any "use" statement

### ISSUE THREAD

**Issue by flaviovs** [vimeo/psalm#6065]
https://psalm.dev/r/b6e56c987d

Comment out the "use" statement on line 6 and the error goes away.

Swapping `use` and doc-block positions also gets rid of the error, but this is far from ideal since the doc-block is supposed to have file-wide documentation (this construct is very common in frameworks that use PHP as template variable, such as Yii).

**Comment by weirdan** [comment]
While this shouldn't affect scalars, `use` position does matter, as it only affects the code that follows: https://3v4l.org/vMriL

So if you reference any imported classes in your docblock you would have to put it *after* the `use`.

**Comment by flaviovs** [comment]
Sure, with the addendum for fully qualified class names, which should also work even before `use`, otherwise file-level doc-blocks would be impossible.

**Comment by kkmuffme** [comment]
By adding "use" statements, PHP assumes the code is within a namespace like: https://psalm.dev/r/8baf37b2e7
And there cannot be any code before the namespace, as that would be fatal 3v4l.org/nKlP7#v8.5.3

Assuming the @var is only used to annotate templates which have the variables inherited from the file that requires the current file:

>otherwise file-level doc-blocks would be impossible.
>with the addendum for fully qualified class names, which should also work even before use

Assume [psalm.dev/r/d2e4530d3f](https://psalm.dev/r/f2665cef70) - $foo is available in both namespaces (when $foo is defined in the parent file that included the current file, otherwise obviously a fatal)

This is ambigous - since B refers to completely different classes in each case.

EDIT: is there any advantage of putting the @var in the docblock above the namespace/use compared to putting it in a comment after it?

**Comment by flaviovs** [comment]
> And there cannot be any code before the namespace, as that would be fatal 3v4l.org/nKlP7#v8.5.3

Documentation isn't code.

> Assuming the @var is only used to annotate templates which have the variables inherited from the file that requires the current file:

That's exactly the use case where I found this issue.

> This is ambigous - since B refers to completely different classes in each case.

Partially agree.

Intuitively (IMHO) it would behaves like `/** @var \B $foo */` so the variable could never be an instance of any of those two _B_. 

In fact you can test this by defining `$foo` as anything not `A\B` or `C\B` in a `main.php` file which `include`s your snippet, and running the former: nothing is echoed.

In other words, `/** @var B $foo */` at the top  _could_ imply the global namespace IMHO, since none has been declared yet.

Now the problem I guess is the ambiguity of `@var`. Ideally those file-level globals should be documented with [@global](https://docs.phpdoc.org/guide/references/phpdoc/tags/global.html), where it would make sense (IMHO) to imply the global namespace for non namespaced types, pretty much as PHP does. However AFAIK that tag isn't supported by Psalm.

> is there any advantage of putting the @var in the docblock above the namespace/use compared to putting it in a comment after it?

From a documentation perspective, yes, since those are truly file-level objects. Depending on how many `use` statements you have, global object documentation gets buried down in the file, which is bad.

**Comment by kkmuffme** [comment]
>Documentation isn't code.

Yes, but @var is documentation that this variable was defined "above" this docblock when used with a variable. Which is not possible, except in case of a parent file.

>Intuitively (IMHO) it would behaves like /** @var \B $foo */ so the variable could never be an instance of any of those two B.

This is also how it behaves if there is no namespace in a file - but since there is/are namespaces, PHP requires all code in this file to be inside a namespace.
While this may sound a bit theoretical, this is what the biggest majority of stubs files look like: they contain often 100s of different namespaces.

>In other words, /** @var B $foo */ at the top could imply the global namespace IMHO, since none has been declared yet.

but what about your previous requirement then:
>with the addendum for fully qualified class names, which should also work even before use

?

>Now the problem I guess is the ambiguity of @var. 

@var bool $a is just a docblock replacement for variable assignments like $a = false; that happened somewhere or somehow in a way that is opaque to psalm. Ideally, you wouldn't need any @var (except for class properties), since psalm is/should be able to infer all types - which is why there's even a config option for it https://psalm.dev/docs/running_psalm/configuration/#disablevarparsing

Since most code-bases have some "auto-magic", this often isn't realistic though - even though for many uses cases, I'd argue it's faster and more reliable to create a psalm plugin that handles that "magic" transparently for psalm, than having to maintain a plethora of @var in lots of files.

>Ideally those file-level globals should be documented with [@global](https://docs.phpdoc.org/guide/references/phpdoc/tags/global.html), where it would make sense (IMHO) to imply the global namespace for non namespaced types, pretty much as PHP does. However AFAIK that tag isn't supported by Psalm.

@global is already supported by psalm - to annotate global variables in a function/method (and soon with https://github.com/vimeo/psalm/pull/11714 also in template files that are required/included bei either of them) which are **explicitly** retrieved/set in **global** context using the global keyword.


Using @global here for variables you pass to a template file from inside a function would be wrong - since these aren't global (and by using them with the `global` keyword in code, you would accidentally make them actual globals, instead of just having them available in the current functions/methods context and included file - which is why the `global` keyword and `@global` tag should be avoided whenever possible)

>From a documentation perspective, yes, since those are truly file-level objects. Depending on how many use statements you have, global object documentation gets buried down in the file, which is bad.

On the other hand, this ensures that you have a distinct docblock comment for inherited/parent declared variables only.
While the top-most comment often includes lots of tags like @copyright, @since, @license, @package and in various cases @author, @see, possibly @deprecated and a description, a separate, distinct docblock would make this clearer.

Also, again, this only is "file-level", if you have only 1 namespace in the file.

---

Summarizing: I totally get your point, especially with templates, and how common of a use case this is (not only WP, but various templating engines use that in a similar way).

If someone is willing to implement this, I'm all for it.

It just has to be limited to files that contain no or just a single namespace to avoid ambiguity.

**Comment by kkmuffme** [comment]
Why not just replace the @var with assert() calls? 
This will ensure that PHP automatically checks that the variable is declared in runtime making it safe if it's changed externally, will allow psalm to detect the type and will not run in production (= no performance penalty) PHP, since assert() statements are noop/ignored there (see the php.ini section on assert for that)
